src/wallet/sqlite.cpp

```diff
@@ -57,8 +57,53 @@ SQLiteDatabase::~SQLiteDatabase()
     }
 }
 
-void SQLiteDatabase::Open(const char* pszMode)
+void SQLiteDatabase::Open(const char* mode)
 {
+    const bool read_only = (!strchr(mode, '+') && !strchr(mode, 'w'));
+
+    const bool create = strchr(mode, 'c') != nullptr;
+    int flags;
+    if (read_only) {
+        flags = SQLITE_OPEN_READONLY;
+    } else {
+        flags = SQLITE_OPEN_READWRITE;
+    }
+    if (create) {
+        flags |= SQLITE_OPEN_CREATE;
+    }
+    if (m_mock) {
+        flags |= SQLITE_OPEN_MEMORY; // In memory database for mock db
+    }
+
+    if (m_db == nullptr) {
+        sqlite3* db = nullptr;
+        int ret = sqlite3_open_v2(m_file_path.c_str(), &db, flags, nullptr);
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to open database: %s\n", sqlite3_errstr(ret)));
+        }
+        // TODO: Maybe(?) Check the file wasn't copied and a duplicate opened
+
+        if (create) {
+            // Make the table for our key-value pairs
+            std::string create_stmt = "CREATE TABLE IF NOT EXISTS main(key BLOB PRIMARY KEY, value BLOB)";
```

Comment:
Just noticed @ryanofsky's comment: https://github.com/bitcoin/bitcoin/pull/19077#discussion_r498437365

I think it's safer to explicitly check if the table is present. In general it's nice to have a clearly separate code path for stuff we only do on wallet creation (and upgrade).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19077#discussion_r499643317

> Just noticed @ryanofsky's comment: [#19077 (comment)](https://github.com/bitcoin/bitcoin/pull/19077#discussion_r498437365)
> 
> I think it's safer to explicitly check if the table is present. In general it's nice to have a clearly separate code path for stuff we only do on wallet creation (and upgrade).

Dropping 'if not exists' and adding an error sounds good (though it might be helpful to say what is unsafe or what dangerous cases you are thinking of). I only suggested using 'if not exists' SQL syntax to avoid reimplementing it in C++.

Reply:
>  Alternatively, if `create` is meant as "create if needed", then it's seems safer to explicitly check if the table doesn't exist yet. In that case the error "Failed to create new database" is incorrect.

It's is use as a "create if needed." All wallet DBs are created with the mode "cr+". I don't think we should be giving an error if the table doesn't exist in that case either.